### PR TITLE
RWA-1725 optimisitic retry mechanism for message processing

### PR DESCRIPTION
**Pull request by vgulapally**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RWA-1725

### Change description ###
Optimistic retry mechansim

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/clients/RestExceptionCategory.java [hunk 1]
```diff
@@ -3,40 +3,19 @@
...
+    private static final List<HttpStatus> NOT_PROCESSABLE_HTTP_ERRORS =
+            List.of(BAD_REQUEST, INTERNAL_SERVER_ERROR);
 
     private RestExceptionCategory() {
-
     }
 
     public static boolean isRetryableError(HttpStatus httpStatus) {
-        return RETRYABLE_ERRORS.contains(httpStatus);
+        return !NOT_PROCESSABLE_HTTP_ERRORS.contains(httpStatus);
```
Comment (raghera) [inline src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/clients/RestExceptionCategory.java:19]: Usually reads better if you avoid double negation.  Perhaps then this method should be called isNonRetryableHttpError or isUnprocessableHttpError then have

`return UNPROCESSABLE_HTTP_ERRORS.contains(httpStatus);`

You'd probably have to change the client code to fit with that.

File: src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/clients/DatabaseMessageConsumer.java [hunk 2]
```diff
@@ -159,24 +159,24 @@ private Optional<MessageUpdateRetry> processMessage(CaseEventMessage caseEventMe
...
 
     private Optional<MessageUpdateRetry> processException(FeignException fce, CaseEventMessage caseEventMessage) {
-        boolean retryableError = false;
+        boolean isNonRetryableError = true;
         try {
             final HttpStatus httpStatus = HttpStatus.valueOf(fce.status());
-            retryableError = RestExceptionCategory.isRetryableError(httpStatus);
+            isNonRetryableError = UnprocessableHttpErrors.isNonRetryableError(httpStatus);
         } catch (IllegalArgumentException iae) {
             if (fce instanceof RetryableException) {
```
Comment (raghera) [inline src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/clients/DatabaseMessageConsumer.java:167]: I know you didn't add this but can an IllegalArgumentException be a RetryableException?  Seems not.

**Review by raghera — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/clients/DatabaseMessageConsumer.java (hunk 2 above)
Comment (vgulapally) [reply src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/clients/DatabaseMessageConsumer.java:167]: yes this is existing piece of code, was also wondering why we have added it so left as is for now

File: src/integrationTest/java/uk/gov/hmcts/reform/wacaseeventhandler/MessageProcessorTest.java [hunk 3]
```diff
@@ -194,7 +194,7 @@ void should_update_hold_until_and_retry_count_for_ready_messages_when_retryable_
                     "classpath:sql/insert_case_event_messages_for_processing_ready_msgs.sql"})
```
Comment (BenNzewi) [inline src/integrationTest/java/uk/gov/hmcts/reform/wacaseeventhandler/MessageProcessorTest.java:194]: do we need to also cover setting message_state to Ready when retryable_error?

Comment (vgulapally) [reply src/integrationTest/java/uk/gov/hmcts/reform/wacaseeventhandler/MessageProcessorTest.java:194]: Ben I have not modified existing logic setting message_state to Ready, have only classified what error are retryable and unprocessable.

**Pull request merged**